### ISSUE TITLE
When will you release version 4?

### ISSUE THREAD

**Issue by EECOLOR** [cssnano/cssnano#408]
It took me a while to figure out that the site with documentation is talking about version 4 as having safe defaults. The instructions just tells to `npm install cssnano --save-dev` which installs version 3 apparently.

So two questions:
1. When will 4 be released?
2. Where can I find the documentation for 3?

Thank you for your hard work.

**Comment by andyjansson** [comment]
>When will 4 be released?

Well, we're currently waiting for the following dependency to be updated: https://github.com/stylelint/stylelint/issues/275

>Where can I find the documentation for 3?

I'm.. not sure 😅 The main difference between 3.x and 4.x is the safe default preset, as you've discovered. 3.x has more or less the same options, but passed directly to cssnano.

**Comment by EECOLOR** [comment]
The stylelint pull request you point to has been merged and released. Did you mean another issue?

I am asking because cssnano is the only thing in my build holding on to postcss 5 (https://github.com/postcss/autoprefixer/issues/905)

**Comment by andyjansson** [comment]
Sorry, I meant the following issue: https://github.com/stylelint/stylelint/issues/2757 

I accidentally removed the last number.

**Comment by EECOLOR** [comment]
For completeness: the stylelint issue is waiting for bugs in https://github.com/postcss/postcss-selector-parser to be fixed

**Comment by EECOLOR** [comment]
Which seems to be stalled because @ben-eb does not have time. Since this (the cssnano) repository is hosted on the account of @ben-eb maybe other contributors of cssnano have enough experience to help fix the issues in `postcss-selector-parser` that are preventing the release of stylelint.

**Comment by caraya** [comment]
You can always install cssnano to point to a specific version or to a Github repository and keep it pinned to that version.

For example: 

`npm install cssnano@next`  will install 4.0.0rc2 or whatever version is tagged with `next` at the time you install

Would that solve your issue until 4.0

**Comment by EECOLOR** [comment]
@caraya It would certainly do, but at my company we can not use release candidates in production software.

**Comment by evilebottnawi** [comment]
@ben-eb friendly ping, it is allow to upgrade `postcss` in `css-loader` and speed up

**Comment by pascalduez** [comment]
Hi,

I have to agree on the documentation point, not being able to consult the current stable version docs is quite disturbing and time consuming.
For instance `css-loader` use the curent 3.x versions.

**Comment by EECOLOR** [comment]
Just an update for anyone interested.

It seems [stylelint/issues/2757](https://github.com/stylelint/stylelint/issues/2757) is at a standstill. Stylelint exposes a few issues with the [postcss/postcss-selector-parser](https://github.com/postcss/postcss-selector-parser) that show the parser is not compliant with the specs.

The guys at stylelint are very willing to help and even work around the problems with the parser. That however would just be a band aid as the postcss-selector-parser (by @ben-eb) is itself also used by cssnano. The problem here is that @ben-eb seems to be unavailable. I do not want to judge because I don't know @ben-eb and there may be personal circumstances preventing him from participating.

I hope there is someone in the cssnano team that has contact with him at a personal level and is able to find a way to add collaborators to the postcss-selector-parser.

An alternative might be someone forking the postcss-selector-parser, fixing the problems and have cssnano depend on that version.

In short, there will (or maybe should) be no version 4 of cssnano until the postcss-selector-parser is fixed (compliant to the specs).

**Comment by EECOLOR** [comment]
The issues have been fixed and merged: https://github.com/stylelint/stylelint/pull/2984

At the next release of stylelint it will not be blocking the new version.

Since cssnano itself relies on the postcss-selector-parser I recommend it updates its dependency to the same (fixed) version.

**Comment by LucasIcarus** [comment]
Any news about this issue?

**Comment by LucasIcarus** [comment]
https://github.com/stylelint/stylelint/issues/3012

seems that stylelint will have its 8.3 soon.

**Comment by ben-eb** [comment]
Hi folks,

Yeah as you can probably imagine I've been really busy over the last few months, I've decided to take full-time employment and that has meant so far a long working week. Although my employer is already benefiting from my work on this project I don't see any plans in the short term that will enable me to contribute in a sustainable way (due to the rather large backlog of work I have to fulfil), nor can I promise anything longer term.

I am not 100% confident in the stability of the 4.x codebase, which is why I wanted to land some release candidates and get a feel for what the direction should be. Unfortunately as our main consumer is webpack, I don't think we'll actually get much further feedback until they upgrade, which might mean that future maintainers have to do a few patch/minor versions after 4.0.0 is landed.

I've already handed administrative access over to @chriseppstein but I can't speak for him. @andyjansson knows probably more than most about the codebase, but we haven't spoken in a while due to work pressure. Other recent contributors have been mostly drive-by, you can check the [activity](https://github.com/ben-eb/cssnano/graphs/contributors) for yourself.

About the docs, it's difficult enough voluntarily maintaining a project with millions of downloads and keeping docs up to date for two release lines, so I hope you can cut me some slack... 😞

**Comment by chriseppstein** [comment]
I want to dig into this, but I have my own work deadlines I have to hit at the moment, so I can't devote any significant time to this for another few weeks.

What I could really use help with is getting my arms around what's needed to do to release 4.0.

I'm happy to fix bugs and cut a release, but I need a hit list of open bugs that folks consider a blocker to adopt.

Taking over a code-base this big is pretty hard, especially when you've not been doing any serious development on it. I could really use the community's help testing. At some point, I'll probably just say "eff it" and release. major versions don't auto-upgrade so if there's any major issues, we'll just have to iterate once we find out about them.

**Comment by EECOLOR** [comment]
@ben-eb Thank you for clearing up your position

@ben-eb @chriseppstein For me it would be good enough in the short term to simply make 3.x work with / depend on postcss 6. I have no clue if that is feasible at all.

**Comment by chriseppstein** [comment]
> I have no clue if that is feasible at all.

@EECOLOR Nor do I, if someone wants to see what's involved and report back, it would be helpful.

As a process point, I do not even have the requisite permissions to contribute to older versions of most of the cssnano dependencies from before the monorepo conversion. Moving forward with 4.0 may be the only route if updates to the dependencies are needed.

**Comment by EECOLOR** [comment]
Ok, so now this library needs to update both stylelint and postcss-selector parser and then it can be released?

**Comment by EECOLOR** [comment]
Any news on this?

**Comment by evilebottnawi** [comment]
@ben-eb friendly ping

**Comment by evilebottnawi** [comment]
@chriseppstein @andyjansson friendly ping, in near future we release webpack@4 and upgrade css-loader, we need first stable release for starting reporting and fixing bug, also you can invite me and show roadmap. I help with fixing bugs :+1:

**Comment by EECOLOR** [comment]
https://github.com/postcss/postcss-selector-parser/pull/133

**Comment by evilebottnawi** [comment]
@chriseppstein friendly ping, we need some fixes and stable release :+1: I can help maintain this package

**Comment by chriseppstein** [comment]
@evilebottnawi @EECOLOR I'm like a week or two from releasing css-blocks, which has been consuming 95% of my open source development time. I'm not great at context switching so I try to keep it to a minimum. I thought I could get the selector parser 3.0 out, but that has already spiraled into more work than I can justify at the moment. I'm going to put a bow on my current changes there and cut another pre-release. If there's other changes needed, I could use a hand in the form of Issues, Pull Requests, and testing against important downstream consumers. In terms of cssnano, I don't have the cycles at the moment to do much more than merge pull requests and cut releases, if one of you wants to take point on getting it into releasable state for 4.0, that would be helpful. Once I see movement on that, I'll update repo access permissions as warranted.

**Comment by EECOLOR** [comment]
@chriseppstein Thank you very much for the effort you have put in so far!

CSS is on the other end of the spectrum for me. I'm mainly a backend developer and we have developed a zero configuration wrapper around webpack. We would like to keep stuff up-to-date and cssnano is the part that holds us back from the new postcss. I don't know a lot about CSS, `display: none;` is about the best I can do haha. What I mean to say is that I don't feel I can help out here.

I really hope that more people with a better skill-set can jump in and help cssnano to 4.0!

PS. I like your css-blocks idea!

**Comment by chriseppstein** [comment]
I've started a project to track issues that need to be addressed to release 4.0

https://github.com/ben-eb/cssnano/projects/1

I really could use a hand with pull requests and triaging issues that are considered blockers or high-priority regressions in 4.0 compared to 3.x.

If there's other issues that aren't yet on that project board, please mention them here and I'll add them.

**Comment by EECOLOR** [comment]
@chriseppstein It might be an idea to add the request for help to the readme, the npm readme and the documentation site.

**Comment by evilebottnawi** [comment]
@chriseppstein ping, i send 3 PR to solve problems and no answer. Can you solve the problem with your time? If you don't have time on project transfer it to people who have at this time, or at least give the right to continue development.
@ai I am very concerned about this situation, maybe somebody from evil martians have permissions and can help with this

**Comment by ai** [comment]
@evilebottnawi yeap, I am very concerned about cssnano 4 as well.

@chriseppstein can we have PR reviews? It is not a big problem to send new developers to fix issues, but without PR review process we will need to find different solution :(.

**Comment by andyjansson** [comment]
>ping, i send 3 PR to solve problems and no answer. 

@evilebottnawi I've merged two of your PRs and left some comments on the third.

**Comment by evilebottnawi** [comment]
@ai what about fork and release on new name? i know it is very very bad, but i send around 10 PR (https://github.com/ben-eb/cssnano/pulls) and all are unanswered/unreviewed, it is impossible to fix problem(s) and release new versions :disappointed: Maybe we can contact with @ben-eb and solve this problem, i don't have @ben-eb contacts

**Comment by ai** [comment]
@evilebottnawi Ben Briggs burned out and left open source.

@andyjansson can we review other PR?

**Comment by andyjansson** [comment]
@ai Of course, but I'm a bit busy until tomorrow. I will get around to it, don't worry.

**Comment by andyjansson** [comment]
Sorry that the amount of my time I donate to this project is insufficient for you. I have a life aside of this, and sometimes I might not be available for a few days.

**Comment by evilebottnawi** [comment]
@andyjansson I know about all people have life aside. I just suggest take care of this project as main contributors. I will do PR, if no unanswered 72 hours, i merge as is (good practice for open source projects which don't have enough contributors). Also take care about all issue bug/features/regressions/documentation and etc. I just can not understand why this is avoided. I offer my help free.

**Comment by bjankord** [comment]
Looks like per #476 that @evilebottnawi will be picking up the project as a maintainer. Thank you for picking this project up! 

Does this get v4 stable closer to a release date? It looks like @evilebottnawi has been working on getting things ready and I saw [the milestone for v4](https://github.com/cssnano/cssnano/milestone/6). Is that milestone / issue list accurate?

**Comment by evilebottnawi** [comment]
@bjankord thanks. I want to release this on this week, but don't have enough time :disappointed: 
Prioritized issue https://github.com/cssnano/cssnano/issues/439, feel free to go ahead and send a PR with fix :+1:

**Comment by ai** [comment]
@evilebottnawi any blockers? do you need help with something?

**Comment by evilebottnawi** [comment]
@ai sorry, a lot of work, only one bloker https://github.com/cssnano/cssnano/issues/439

**Comment by bjankord** [comment]
Can that issue go in a 4.1.0 or 4.0.1 release to unblock the 4.0.0 release?

**Comment by evilebottnawi** [comment]
@bjankord better fix this in `4.0.0`, it should be not difficult, just don't have time right now, in todo list for tomorrow

**Comment by evilebottnawi** [comment]
@ai can't release `4.0.0`, because don't have access to many repos in `npm`: css-size, cssnano-preset-advanced, cssnano-preset-default, cssnano-util-get-arguments, cssnano-util-get-match and a lot of other.
Solution:
1. Looks you have access, clone repo and run `lerna publish --skip-git`, then `git tag 4.0.0 and git push` and add release https://github.com/cssnano/cssnano/releases
2. Give me access to repos.

**Comment by ai** [comment]
@evilebottnawi I gave you a access for all packages in `packages/`. 4.0!!! 🎉

**Comment by jonnitto** [comment]
If I run my script with version 4, I got following error: 
```
BrowserslistError: Unknown version 67 of android
```
If a remove cssnano or revert to the old version, everything works fine.

**Comment by evilebottnawi** [comment]
@jonnitto looks you have bad `browserslist` file, can you put here

**Comment by jonnitto** [comment]
@evilebottnawi I've updated my comment above

**Comment by evilebottnawi** [comment]
@jonnitto can't reproduce, can you try `rm -r package.lock && rm -r node_modules && npm cache clear --force && npm i`

**Comment by jonnitto** [comment]
@evilebottnawi I created a build package and I found a way to solve this problem. Thank you for your hint! it didn't solve it but gave me an idea. 

**Another problem**
If I have some css like this:
```css
body {
    min-height: calc(100vh - constant(safe-area-inset-top));
    min-height: calc(100vh - env(safe-area-inset-top));
}
```
I receive following error:

```bash
Parse error on line 1:
100vh - constant(safe-area-i
--------^
Expecting 'SUB', 'LPAREN', 'NESTED_CALC', 'NUMBER', 'CSS_VAR', 'LENGTH', 'ANGLE', 'TIME', 'FREQ', 'RES', 'EMS', 'EXS', 'CHS', 'REMS', 'VHS', 'VWS', 'VMINS', 'VMAXS', 'PERCENTAGE', got 'PREFIX'
```

**Comment by andyjansson** [comment]
@jonnitto I wrote the current implementation of `postcss-calc`. As @evilebottnawi says, it only works on valid CSS. You need to transform your CSS, resolving your `constant` and `env` functions before `postcss-calc` executes.

**Comment by jonnitto** [comment]
Thank you for the hint. I could solve it with first creating a CSS property and then using this property in `calc`. Thank you both a lot for your great work ❤️

**Comment by evilebottnawi** [comment]
Please open new issue, don't write here if you have problem after upgrade.
